sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
This will also trigger a copy of `init` even if `init` is already an array with the right dtype while this was not the case before.

I think this is fine to systematically copy as it is probably a bad idea to mutate the user provided input array silently. I think you should add a test to check this behavior.